tests/realtikvtest/txntest/stale_read_test.go

```diff
@@ -1506,3 +1493,245 @@ func TestStaleReadNoBackoff(t *testing.T) {
 	require.Regexp(t, ".*rpc_errors:{data_is_not_ready:1.*", explain)
 	require.NotRegexp(t, ".*dataNotReady_backoff.*", explain)
 }
+
+func TestStaleReadAllCombinations(t *testing.T) {
+	store := realtikvtest.CreateMockStoreAndSetup(t)
+	defer config.RestoreFunc()()
+
+	tk := testkit.NewTestKit(t, store)
+
+	safePointName := "tikv_gc_safe_point"
+	safePointValue := "20160102-15:04:05 -0700"
+	safePointComment := "All versions after safe point can be accessed. (DO NOT EDIT)"
+	updateSafePoint := fmt.Sprintf(`INSERT INTO mysql.tidb VALUES ('%[1]s', '%[2]s', '%[3]s')
+	ON DUPLICATE KEY
+	UPDATE variable_value = '%[2]s', comment = '%[3]s'`, safePointName, safePointValue, safePointComment)
+	tk.MustExec(updateSafePoint)
+
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t (id int primary key, v int)")
+
+	// Insert row #1
+	tk.MustExec("insert into t values (1, 10)")
+	time.Sleep(1000 * time.Millisecond)
+	firstTime := time.Now().Add(-500 * time.Millisecond)
+	// Retrieve current TSO from store's Oracle instead of @@tidb_current_ts.
+	externalTS, err := store.GetOracle().GetTimestamp(context.Background(), &oracle.Option{})
+	if err != nil {
+		t.Fatalf("failed to get TSO: %v", err)
+	}
+
+	time.Sleep(1000 * time.Millisecond)
+	// Insert row #2
+	tk.MustExec("insert into t values (2, 20)")
+	time.Sleep(1000 * time.Millisecond)
+	secondTime := time.Now().Add(-500 * time.Millisecond)
+
+	staleReadMethods := []struct {
+		name   string
+		setup  func()
+		query  string
+		clean  func()
+		expect []string
+	}{
+		{
+			name: "tidb_read_staleness",
+			setup: func() {
+				tk.MustExec("set @@tidb_read_staleness='-2'")
+			},
+			query: "select * from t",
+			clean: func() {
+				tk.MustExec("set @@tidb_read_staleness=''")
+			},
+			expect: []string{"1 10"},
+		},
+		{
+			name:   "AS OF TIMESTAMP sees only first row",
+			setup:  func() {},
+			query:  fmt.Sprintf("SELECT * FROM t AS OF TIMESTAMP '%s'", firstTime.Format("2006-1-2 15:04:05.000")),
+			clean:  func() {},
+			expect: []string{"1 10"},
+		},
+		{
+			name:   "AS OF TIMESTAMP sees both rows (using secondTime)",
+			setup:  func() {},
+			query:  fmt.Sprintf("SELECT * FROM t AS OF TIMESTAMP '%s'", secondTime.Format("2006-1-2 15:04:05.000")),
+			clean:  func() {},
+			expect: []string{"1 10", "2 20"},
+		},
+		/*
+			We cannot test TIDB_BOUNDED_STALENESS here because it relies
+			on the update of safe ts, which cannot be controlled by the test
```

Comment:
🤔 Do we have any other way to test this case?

Reply:
I haven't found an effective way to implement this in the combinatorial test.